### PR TITLE
Refactorization and significant improvements in efficiency, plus a handful of new features

### PR DESCRIPTION
**Pull request by ThomasLastName**
The BNN class has been torn apart and split into several pieces, each of which are now their own class.

Class corresponds to the choice of prior and variational family.

The only real variational family that's been implemented is `IndepLocScaleBNN`, which does not come equipped with any prior distribution. With the exception of the base class `BayesianModule`, all other classes sub-class `IndepLocScaleBNN` and add the machinery necessary for some particular prior distribution.

Most significant efficiency improvements are the parallelization of the `forward` and `prior_forward` methods in the BNN and Ensemble classes. The key was to use `torch.bmm`.

### CODE REVIEW

**Comment by natalieklein229** [comment]
Hi @ThomasLastName 

Here are a few notes as I look through the code. I am generally OK with merging now and addressing some of these later, perhaps, and some of them are more stylistic than anything — so not crucial to make changes. I also realize you have limited time right now so, again, if easier to just merge in now and perhaps we look at some of these things later, that’s fine with me. 

Experiments
- In experiments, for example `bnns/experiments/train_bnn.py`: the way you put hyperparameters into globals does trigger a lot of squiggly lines in VSCode because it can’t resolve the variables. I’m not sure global variables are the way to go. The typical I’ve seen in many scripts is to use argparse to get a namespace, and then reference everything as `args.device` for instance. 
- Also, you may have a good reason for this, but when you instantiate BNN why not pass many of the hyper parameters as additional arguments instead of then having to set them manually just afterward? There's risk that a user may not set everything needed and there may not be a default value set, causing errors.
- Why is `predict()` defined as a local function in the experiment scripts rather than a method of BNN?
- How hard would it be to remove the dependency on `quality_of_life` by copying the relevant functions into `util`? This would be cleaner for a user installing things, and we wouldn’t have to worry about you changing `quality_of_life` and possibly breaking something. 

Small style thing 
- Some of your lines are very long which can be hard to read. There are utilities like `black` that can help auto format code a bit, but you’d have to see what it does (and there might be better options).

Models
- `univar_BNN/` and related code; a lot of the code is redundant, seems like you could make a function that can construct the `SequentialGaussianBNN` based on some passed in layer widths, and an argument for dropout, for instance.

Scripts
- There is a lot going on here. `bnns/experiments/` I guess are the main entry point for users, but it might be worth considering cleaning up and/or documenting the `scripts` a bit.

**Comment by ThomasLastName** [comment]
Thanks for looking at this! I tend to agree with these suggestions. Especially, I agree that the training scripts (most notably `train_bnn.py`) are just about the messiest part of the codebase and should be clean up.

Experiments:
- I agree it should be cleaned up. Of course, the goal is to avoid passing a million different arguments like `python train_bnn.py --lr 0.01 --width 100 --depth 3 -- activation "ReLU" ...`. I'm not familiar with the `args.device` approach you mention, specifically. The current API loads a `.json` file as a dictionary `h_dict` and then dumps its contents into the global scope of the training script. What about just referencing `h_dict["MODEL"]` instead of the global variable `MODEL`, and so on? **Is this equivalent to what you're suggesting?**

- So, why not pass the hyper-parameters as arguments in the `__init__` method of the BNN classes? My initial concern was that, in that case, the training script might require additional complexity in order to figure out "which are the arguments that I should be passing to this class." However, in hindsight, I'm realizing that this complexity just ended up somewhere else: in the `set_prior_hyperparameters` method of BNN classes (which is defined in [WeightPriorBNNs.py](https://github.com/ThomasLastName/multiverse/blob/main/functional_bnns/bnns/WeightPriorBNNs.py) and [GPPriorBNNs.py](https://github.com/ThomasLastName/multiverse/blob/main/functional_bnns/bnns/GPPriorBNNs.py)). I have no objections about migrating that logic to `train_bnn.py` (though it does mean adding to the currently messiest file), and instead making the prior hyperparameters keyword arguments, like how `likelihood_std` is already handled. **What do you think about "migrating that logic" as described, and just getting rid of the `set_prior_hyperparameters` method entirely?**

- The idea was that `predict()` may depend on *both* the data and model, which don't come together until the training script. For instance, with the SLOSH data, predicting the heatmaps requires more than just the forward pass of the model; to recover the actual predicted heatmap, you need to take a linear combination of the "principal heat maps." So, essentially, `predict()` handles such "downstream" tasks. However, since the implementation isn't fully generally anyway... maybe it's better to fall back to a simple if/else (e.g., `if SLOSH_DATA: post_process_to_get_predicted_heatmaps`)? In the opposite extreme, a fully general implementation would be to add `PREDICT` as another config setting and load that function from a user-specified path. **Do you want to go more general, or simpler?** in this situation, personally, I advocate for simpler. In any case, I need to update the `predict()` functions to leverage the new `n` keyword of the BNNs' forward methods. Another purpose of, `predict()` is to handle dataloaders. However, now I'm thinking maybe I'll just add this as a feature in the next update:
```
def predict( self, x, n=1 ):
    if x is a dataloader:
        return stack(
            self.predict(batch,n) # ~~~ current implementation
            for batch in x
        )
    else:
        return the result of the current implementation
 ```

- It would be easy to remove the dependency on `quality_of_life`. It would amount to copy-pasting a few hundred extra lines into the `utils.py` file. **I'll do that in the next update.** P.S. That file's getting pretty long. If you don't object, I'd like to split it into several files (e.g., `math_utils.py`, `plotting_utils.py`, etc.) by having a `utils` folder containing multiple `.py` files. I think this can be done without needing to rewrite import statements.

Style:
- I recall Michael mentioning such tools! I'm interested, however I fear it would be harder to adopt them now, since we didn't make a point of working with them from the start... I should give the whole codebase a final read-through for quality assurance before concluding the project. That might be a good time to tweak style. Though I can't promise to implement them algorithmically, **do you have any other suggestions in addition to more line breaks?** (other than adding docstrings, the absence of which I know is egregious... it's at least on my todo list, and hopefully an easy application of chat-GPT + proof reading).

Models:
- As a matter of fact, the folder `univar_BNNs/` is no longer in user as of [Jan/18](https://github.com/ThomasLastName/multiverse/commit/417b64bbc2e2bea8ed2af5f1be79a2a50d55ab2c) when `MODEL` was replaced by the combination of `MODEL` and `ARCHITECTURE`. I guess you could call it deprecated, but I just didn't feel the need to delete it. Maybe a deprecation warning is overkill and I should just delete this folder since the package isn't in use yet (to my knowledge)
Screenshot 2025-05-23 090238

 - Complicating matters further, the `univar_NN_dropout/` folder is similarly deprecated but neglects to throw a deprecation warning (my oversight), while `univar_NN/` remains in use. I suggest deleting these two folders, but keeping the folder `univar_NN/`. If so, then later I'd plan to add a `slosh_NN/` folder in a simliar fashion.
  - In principle, it may be possible to specify `LAYER = [30, 30, 30]` and `ACTIVATION = "ReLU"` instead of `ARCHITECTURE = "univar_NN.univar_NN_30_30_30"`, if you know what I mean. However, the motivation behind loading a model from a user-specified path is to allow for generality. For instance, you can store `[30, 30, 30]` as a field in a `.json` file, but you couldn't store `[ 30, 30, 30, nn.Softmax ]`. Instead, the current solution is to define `ARCHITECTURE = "file_name"` where you define the actual architecture in `models/file_name.py`. Note that the models defined in `univar_NN/` have an `nn.Unflatten` layer, whereas we wouldn't want this for the SLOSH data. It's those kind of cases that I find difficult to handle with a simple function that builds a neural net with specified number and width of layers. You could reduce the file-count by instead having a file `models/network_builder.py` where you define a function `build_relu_network( *layers, activation )`. Then, instead of `ARCHITECTURE = "univar_NN.univar_NN_30_30_30"` perhaps you could specify `LAYER = [30, 30, 30]` and `NETWORK_BUILDER = "build_relu_network"`. However, again, it seems difficult to account for possible future use-cases like convolutional layers. For the moment, **I advocate to leave `univar_NN/` as it is currently, while deleting `univar_BNN/` and `univar_NN_dropout/`**. However, I'm open to more discussion on this matter, too. If we keep `univar_NN/` then, in the near future, I'd add a similar folder `slosh_NN/`.

Thank you for the continued feedback, @natalieklein229 !

**Comment by ThomasLastName** [comment]
P.S. If we're on the same page, I can cancel this pull request, implement most/all of these as quick fixes, and then re-submit the pull request after a week or so.

**Comment by natalieklein229** [comment]
Hi @ThomasLastName 

Experiments

Argument passing: Yes, I would prefer you reference the dictionary keys instead of dumping into global variables, as global variables can be tricky, hard to track, accidentally overwritten, etc. The args.device I was referring to is what you get if you use the argparse method to get command line arguments, but it’s not exactly compatible with JSON so I think just keeping as a dictionary is fine. For future reference, I do think argparse is by far the most common way I’ve seen ML folks handle hyperparameters, and you can also use config files with it, similar to https://micha-feigin.medium.com/on-using-config-files-with-pythons-argparse-8af09d0bdfb9 but I don’t think it’s essential to switch over, more essential to not have the global variables.

Hyperparameters in init: I think it would be cleaner if you provide ability to pass everything during init including prior hyperparameters. I would provide reasonable default values for all of those in the init function so that it doesn’t necessarily have to be messy (if using the defaults). 

Predict: I think it’s okay the way you have it. Predict is a method of nn.module in PyTorch though; the whole idea there is that it basically just does the forward on a batch. Having custom “predict” functions that also do special other things depending on data is fine.

Util: sounds good to me!

Style

You could experiment with `black` or an alternative just to see what it does on one file, then could always revert with git checkout if you don’t like what it does. When I’ve used it, it’s primarily adding line breaks to make it more readable. I can’t recall if it gives you docstring stubs to fill in. I don’t really have other suggestions right now.

Models

I would advocate deleting models that are no longer in use. You can keep a backup of the code locally or in a different fork if you want, just in case, but no, I don’t think you will be disrupting any current use of the code that I know of. Unfortunately, our funding ended so the postdoc I did have working on some things didn’t get that far with your code and isn’t currently working on anything related.

You can leave the other models as is, I wouldn’t worry too much about `model_builder`.

**Comment by natalieklein229** [comment]
@ThomasLastName I see that you're making changes here, please just let me know when I should review it before merge

**Comment by ThomasLastName** [comment]
Right, I did a little work on it over the weekend. I don't want to draw it out too long, but I wasn't able to address every point quite yet. So, **I'm planning to work on this more before finalizing it.**

Looks like the changes are automatically reflected? I hope you aren't getting a notification for every commit!

**Comment by ThomasLastName** [comment]
I'm close but still not quite done with addressing these points. Please bear with me a little longer.

**Comment by ThomasLastName** [comment]
Alright, this is essentially ready to be merged, except I haven't applied `black` yet.

I downloaded `black` and applied it to few files experimentally, but haven't committed this change yet. It seems fine. I'm happy to use `black` if it would make the code feel more approachable to you, Natalie, or anyone else. I just haven't taken the time to apply it to every single file yet. I'll try to do that in the evening on like Monday or Tuesday (should just talk like <1 hour, I figure?) unless you merge before then, which would also be fine.

**Comment by ThomasLastName** [comment]
Ohhh, I can just `black .` instead of doing one file at a time.

I checked a handful of the files and ran like two tests. Otherwise, I'm just trusting in `black` to not break the code (anyway, worst case, we just rewind 1 commit).

**Comment by ThomasLastName** [comment]
Now would be a good time to merge, if there are no other concerns we want to address first. I'll continue to make changes, but this is the majority.

A couple of misc notes:
 - I think I forgot to respond to this question, but the `scripts/` folder is honestly just my personal workspace. Some of those scripts could justifiably be moved to the `experiments/` folder. Those which can't are simplistic tests.
 - Speaking of global variables in the training scripts, note that VS code is similarly unable parse the functions I load using `from metrics import *`
